akka-actor-typed-tests/src/test/scala/akka/actor/typed/internal/delivery/ReliableDeliveryShardingSpec.scala

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (C) 2019-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.delivery
+
+import scala.concurrent.duration._
+
+import akka.Done
+import akka.actor.testkit.typed.scaladsl.LogCapturing
+import akka.actor.testkit.typed.scaladsl.ScalaTestWithActorTestKit
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.internal.delivery.ConsumerController.SequencedMessage
+import akka.actor.typed.internal.delivery.SimuatedSharding.ShardingEnvelope
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.LoggerOps
+import org.scalatest.wordspec.AnyWordSpecLike
+
+object ReliableDeliveryShardingSpec {
+
+  object TestShardingProducer {
+
+    trait Command
+    final case class RequestNext(sendToRef: ActorRef[ShardingEnvelope[TestConsumer.Job]]) extends Command
+
+    private final case object Tick extends Command
+
+    def apply(producerController: ActorRef[ShardingProducerController.Start[TestConsumer.Job]]): Behavior[Command] = {
+      Behaviors.setup { context =>
+        context.setLoggerName("TestShardingProducer")
+        val requestNextAdapter: ActorRef[ShardingProducerController.RequestNext[TestConsumer.Job]] =
+          context.messageAdapter(req => RequestNext(req.sendNextTo))
+        producerController ! ShardingProducerController.Start(requestNextAdapter)
+
+        // simulate fast producer
+        Behaviors.withTimers { timers =>
+          timers.startTimerWithFixedDelay(Tick, Tick, 20.millis)
+          idle(0)
+        }
+      }
+    }
+
+    private def idle(n: Int): Behavior[Command] = {
+      Behaviors.receiveMessage {
+        case Tick                => Behaviors.same
+        case RequestNext(sendTo) => active(n + 1, sendTo)
```

Comment:
I'm struggling a bit to see how I'd actually use this

Since this is pulled work, the most common use case of using sharding, like a HTTP endpoint posting one off data for an entity would not be a good fit for reliable delivery since it would have to do a start + wait for requestNext?

Or would you keep the `ShardingProducerController` around, with a buffer and then let all requests go through that?

Perhaps for example consuming a stream and passing elements to sharding would work because then you'd have a single producer repeatedly emitting?

Reply:
Yes, there should be one ShardingProducerController per node/entityType (or producer source). Messages for all entities of that type are sent via that.

For sharding the RequestNext protocol can't be as strict as for plain ProducerController because you should be able to send a new message to a new entity event though there is no demand from the already used entities. Therefore additional demand information is included in the RequestNext. It's even allowed to send to an entity that doesn't have demand and that will be buffered by the ShardingProducerController.

For the HTTP endpoint you would probably use ask via the RequestNext.askNextTo so that you get a reply when it has been confirmed.